### ISSUE TITLE
Feature Request - Tracking weight for the purposes of load capacity

### ISSUE THREAD

**Issue by Nicole-Ridgwell-NMMNHS** [ArctosDB/arctos#5573]
Issue Documentation is http://handbook.arctosdb.org/how_to/How-to-Use-Issues-in-Arctos.html

**Is your feature request related to a problem? Please describe.**
I need to be able to track weights of specimen parts in order to make sure load capacity of shelving is not exceeded.

**Describe what you're trying to accomplish** 
A safe collections environment.

**Describe the solution you'd like**
Ideally, it would be nice to go into the container and see a sum of the weights of all the parts in the container. I'd definitely settle for being able download a csv of all parts in the container that contains part attribute weight. It would be nice to add load capacity to container. As a stop gap for our current project, I can manage with just part attribute weight (see #5572)

**Comment by Jegelewicz** [comment]
> It would be nice to add load capacity to container.

Please file a separate issue - I think this would be a VERY useful bit of information

**Comment by dustymc** [comment]
>container

That's what I was wondering about from the other issue. Herp collections probably don't much care what a pickled salamander weighs, but they might care what a big jar of them weighs.

Can weight (and capacity, while we're in there) just be a NULLable attribute of containers? You might end up with some 10KKG tags (glued to big rocks), the herp collections would end up with a whole lot of NULL tags (which is fine, they're never put directly on shelves) and a few 10KG jars, etc.

That would mean this is only available to those with object tracking data, but I think that's fairly realistic - without machine-readable labels the data are always dicey anyway; some remarks field is adequate.

**Comment by Nicole-Ridgwell-NMMNHS** [comment]
@dustymc I think what you are saying might work. I was trying to figure out how I could track weights of specimens and mounts together and separately (cradles can be pretty hefty), but if we have container cradle with the cradle weight and container tag with the weight of the actual specimen, it will add up properly for the shelf capacity.

**Comment by dustymc** [comment]
>add up properly 

Yup, and without worrying about the details, or needing to catalog everything (so there can be a part to carry the value) to know if you're about to make a big crash, or when you've got a bucket of 86000 unweighed but individually cataloged microtus teeth, or WHATEVER  - weigh the thing you're carrying around, stick it on some shelf with that much remaining capacity, done.

**Comment by Jegelewicz** [comment]
container weight - how much the container weighs
container load capacity - how much the container can hold

add weight units code table

need trigger so you cannot exceed load capacity

**Comment by Jegelewicz** [comment]
@Nicole-Ridgwell-NMMNHS do we need to do anything with this?

**Comment by dustymc** [comment]
I have an administrative question concerning https://github.com/ArctosDB/internal/issues/258: How should an issue get nominated for 'ready for implementation'? I've already flagged it next task, which is probably the start of that process, but now what?

**Comment by Jegelewicz** [comment]
> but now what?

I'd say that you write up exactly what you plan to do. I think that should happen in arctos_dev? AWG reviews and then approves or it gets sent back here for more work/discussion?

**Comment by dustymc** [comment]
I'd rather avoid having any say in what enters the development channel unless that's going to be part of some major change. Sustainability will not allow us to keep abandoning partially-completed things, and I feel like anything that I initiate is going to be blocked.

I've indicated that I think I know what needs to be done and know how to do it with the aforementioned next task milestone. I'm happy to provide any sort of additional information, but the decision needs to be very solid (eg, will happen barring something exposed in the tech-slanted summary in arctos-dev) before things leave this repo if the other is to effectively act as a roadmap/release notes.

https://github.com/ArctosDB/arctos/issues/5573#issuecomment-1462978671 provides an accurate summary.

**Comment by Jegelewicz** [comment]
OK - cool. So you would like us to put this in arctos_dev or just approve it for there here?

**Comment by dustymc** [comment]
No, I need to put things to arctos-dev (that's how you'll know if I actually understand what's been requested or not), just somehow let me know that the decision has been made please.

**Comment by Jegelewicz** [comment]
Actually, are these just code table things that we should file separately? I am looking at container environment and wondering if that, along with these things are container attributes? Should we have ctcontainer_attribute_type that includes

container environment (controlled by [ctcontainer_env_parameter](https://arctos.database.museum/info/ctDocumentation.cfm?table=ctcontainer_env_parameter))
container weight - how much the container weighs (number + units)
container load capacity - how much the container can hold (number + units)

Add other attributes as needed?

Although looking at the container environment values - that seems a little bit messy. Some of those could just become their own attribute - like [checked](https://arctos.database.museum/info/ctDocumentation.cfm?table=ctcontainer_env_parameter#checked)?

I don't want to drag this out any longer than necessary for @Nicole-Ridgwell-NMMNHS who has been patiently waiting fo nearly half a year. So Rather than add the container environment stuff in here, we can deal with that later. Here is what I suggest we do.

Add a new code table ctcontainer_attribute_type

Purpose | Table | Description
-- | -- | -- 
Object Tracking: Container Attribute Types | ctcontainer_attribute_type | Controlled vocabulary for container attribute types.

Add two new terms to the code table

container weight - how much the container weighs (number + weight units)
container load capacity - how much the container can hold (number + weight units)

Add a trigger so that the weight of any containers nested within a container cannot exceed the holding container's load capacity

Alternatively, if the code table thing is plain dumb, then this.

Add two new fields to container

container weight - how much the container weighs (number + weight units)
container load capacity - how much the container can hold (number + weight units)

Add a trigger so that the weight of any containers nested within a container cannot exceed the holding container's load capacity

**Comment by dustymc** [comment]
I'm not sure.

Weight and capacity seem close enough to 'environment' to be justified there.

That table's not really meant to be edited, it's more like a log, so that would be a significant change in both UI and viewpoint.

I'd have to add some sort of maybe-complicated way of preventing random self-conflicting nonsense

* capacity: 1g
* capacity:1kkg

to have any hope of writing a trigger which in turn can make sense of the attributes.

Querying the attributes is more expensive (computationally), and containers do strain our resources.

I think I'd prefer to have these as 'columns' where they can be more controlled, particularly if they're viewed as safety information (and IDK what else they could be), and I think that's probably simpler (one edit form, two-ish bulkloaders, done?) but I'm certainly open to other viewpoints.

**Comment by Jegelewicz** [comment]
Great - then let's go with

Add two new fields to container

container weight - how much the container weighs (number + weight units)
container load capacity - how much the container can hold (number + weight units)

Add a trigger so that the weight of any containers nested within a container cannot exceed the holding container's load capacity

@ArctosDB/arctos-working-group-officers what else needs to get done before Dusty can proceed?

**Comment by dustymc** [comment]
>something like

If you mean arctos-dev, my template is https://github.com/ArctosDB/arctos-dev/issues/new?assignees=&labels=&projects=&template=dev-ready-issue.md&title= and suggestions are most welcome.

If you mean here, I'm not sure I can do better than https://github.com/ArctosDB/arctos/issues/5573#issuecomment-1462978671 but happy to try if you'll let me know what's missing.

**Comment by ewommack** [comment]
> If you mean here, I'm not sure I can do better than https://github.com/ArctosDB/arctos/issues/5573#issuecomment-1462978671 but happy to try if you'll let me know what's missing.

So what you are asking for here in this issue from Arctos Officers is to approve just this request?

Or asking Arctos Officers what steps do we take in any issue that will signal that I can move an issue to Arctos-dev?
If the second one I would say please fill out one of these sections:
**Please see below the notice on what that will be moved to Arctos-dev**
**Summary of work that will be done:**

**Summary of changes to present features in Arctos:**

**Summary of new feature that will be completed by this work:**

Having the summary in bold specifically stated means that there won't be any confusion. Even if you want to copy and paste from a previous comment that would be ok. Having it specifically laid really help follow your process.

**Comment by dustymc** [comment]
>Having it specifically laid really help follow your process.

OK, and if there's some discussion still to be had I'd rather it happen here so that arctos-dev can remain a clear development log. Let's plan for the summary to happen here and then get copied over, which I think means I need to provide a preliminary summary for all 'next task' issues. Reasonable?

--------------------------------


## Source

https://github.com/ArctosDB/arctos/issues/5573

## Summary

Add weight and capacity to table container.

Add trigger to enforce capacity

## User Interface Changes

4 new optional fields in create and edit interfaces


## Data Structure Changes

Add to table container

* weight numeric
* weight_units fkey ctweight_units
* weight_capacity numeric
* weight_capacity_units  fkey ctweight_units

Triggers:

1. Keep values and units paired
2. Prevent container+contents from exceeding load capacity

**Comment by Nicole-Ridgwell-NMMNHS** [comment]
Thanks for figuring out how to move this forward!

**Comment by dustymc** [comment]
This is running at test, exposed in the edit container UI, and has survived very basic testing. Please break it!

Performance isn't great, especially on very deep stacks, but I don't think that will cause actual problems - most container update operations don't particularly need to be very performant, and most institutions probably don't really have weight capacity restrictions.

**Comment by Nicole-Ridgwell-NMMNHS** [comment]
I don't have the link for test

**Comment by dustymc** [comment]
https://arctos-test.tacc.utexas.edu

**Comment by Nicole-Ridgwell-NMMNHS** [comment]
The only test account I could get back into was NICOLEVOLDEN, and it isn't an operator account. Can you make that an operator and add container permissions for NMMNH Paleo?

**Comment by dustymc** [comment]
Not easily, and I'm interested in what happened to your account - I think things probably shouldn't be allowed to get to that state.

Anyway, try

testymctester

``Seekrut.Passwerd1``

and let me know if it needs more acess

**Comment by Nicole-Ridgwell-NMMNHS** [comment]
Thanks! Yeah, I tried several usernames I knew I had used in the past, but other than the one, when I tried to log in, I got "user not found" and password reset said "Sorry, that email wasn't found with your username." Now that I'm in, I can see that username comes up when I search for it in manage operator, but when I click to edit that operator username I get "0 records found for username NICOLERIDGWELL"

Anyway, I was able to overload container NMMNHG000001 easily, it didn't give me any warnings or anything.

**Comment by dustymc** [comment]
> easily

Elaborate please

**Comment by Nicole-Ridgwell-NMMNHS** [comment]
I added a weight capacity (1000 lb) to NMMNHG000001. I had two collection objects in it. I then gave those weights of 900 lbs and 50 lbs. I then went to another collection object in another container, gave it a weight of 100 lbs, and saved. Then I moved that collection object into NMMNHG000001. I repeated that with another collection object that I gave a weight of 1000 lbs, and still no error.  

> move an already-heavy container into an already-limited container

Yes
